flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/JoinUtil.scala

```diff
@@ -0,0 +1,488 @@
+import org.apache.calcite.rex._
+import org.apache.calcite.sql.fun.SqlStdOperatorTable
+import org.apache.calcite.sql.parser.SqlParserPos
+import org.apache.calcite.sql.{SqlIntervalQualifier, SqlKind}
+import org.apache.flink.api.common.functions.FilterFunction
+import org.apache.flink.api.common.typeinfo.{BasicTypeInfo, TypeInformation}
+import org.apache.flink.table.api.{TableConfig, TableException}
+import org.apache.flink.table.calcite.FlinkTypeFactory
+import org.apache.flink.table.codegen.{CodeGenerator, ExpressionReducer}
+import org.apache.flink.table.plan.nodes.logical.FlinkLogicalJoin
+import org.apache.flink.table.runtime.FilterRunner
+import org.apache.flink.types.Row
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable.ArrayBuffer
+
+
+object JoinUtil {
+
+  /**
+    * Analyze join condition to get equi-conditon and other condition
+    * @param  joinNode   logicaljoin node
+    * @param  expression the function to generate condition string
+    */
+  private[flink] def analyzeJoinCondition(
+    joinNode: FlinkLogicalJoin,
+    expression: (RexNode, List[String], Option[List[RexNode]]) => String) = {
+
+    val joinInfo = joinNode.analyzeCondition()
+    val keyPairs = joinInfo.pairs.toList
+    val otherCondition =
+      if(joinInfo.isEqui) null
+      else joinInfo.getRemaining(joinNode.getCluster.getRexBuilder)
+
+    val leftKeys = ArrayBuffer.empty[Int]
+    val rightKeys = ArrayBuffer.empty[Int]
+    if (!keyPairs.isEmpty) {
+      val leftFields = joinNode.getLeft.getRowType.getFieldList
+      val rightFields = joinNode.getRight.getRowType.getFieldList
+
+      keyPairs.foreach(pair => {
+        val leftKeyType = leftFields.get(pair.source).getType.getSqlTypeName
+        val rightKeyType = rightFields.get(pair.target).getType.getSqlTypeName
+
+        // check if keys are compatible
+        if (leftKeyType == rightKeyType) {
+          // add key pair
+          leftKeys.append(pair.source)
+          rightKeys.append(pair.target)
+        } else {
+          throw TableException(
+            "Equality join predicate on incompatible types.\n" +
+              s"\tLeft: ${joinNode.getLeft.toString},\n" +
+              s"\tRight: ${joinNode.getRight.toString},\n" +
+              s"\tCondition: (${expression(joinNode.getCondition,
+                joinNode.getRowType.getFieldNames.toList, None)})"
+          )
+        }
+      })
+    }
+    (leftKeys.toArray, rightKeys.toArray, otherCondition)
+  }
+
+  /**
+    * Analyze time-condtion to get time boundary for each stream and get the time type
+    * and return condition without time-condition.
+    *
+    * @param  condition   other condtion include time-condition
+    * @param  leftFieldCount left stream fields count
+    * @param  inputType   left and right connect stream type
+    * @param  rexBuilder   util to build rexNode
+    * @param  config      table environment config
+    */
+  private[flink] def analyzeTimeBoundary(
```

Comment:
Hi @fhueske, i think this solution is more clearer exactly. I will rewrite the logic asap. But i think we still need to distinguish between greater/smaller at step 5, if the condition is 
"a.proctime > b.proctime - interval '1' second and
b.proctime < a.proctime + interval '1' second"
we also need to throw an exception.
So i want to extract the offset and mark which input the offset belong to, if they belong to the same input, i will throw an exception. 
what do you think?